Add automated pixel pair mitigation for inactive pixel regions

makortel opened the pull request:
This PR adds automated pixel pair mitigation for regions in pixel where there are inactive areas in two layers.

The algorithm has been presented in tracking POG meeting on November 6th
https://indico.cern.ch/event/675097/contributions/2775830/attachments/1553214/2442464/updated_slides_trk_mk_20171106.pdf
A notable addition to the version presented is an option to ignore single inactive panels in FPix from the inactive areas. This option reduces fake rate without notable change in efficiency. Here are MTV plots comparing 1k ttbar+PU50 events in 940pre3 for the pre-release, the "POG meeting version", and the version ignoring single inactive panels
https://cms-tracking-validation.web.cern.ch/cms-tracking-validation/PR/CMSSW_9_4_0_pre3_PR21630/

The algorithm is implemented in `PixelInactiveAreaTrackingRegionsSeedingLayersProducer`, which produces
* `SeedingLayerSetsHits` (normally produced by `SeedingLayersEDProducer`)
   * Technically could be produced with `SeedingLayersEDProducer` here as well, but then I'd need new accessors for some details in `SeedingLayerSetsHits`, and I felt that the current solution would be a bit cleaner
* `TrackingRegionsSeedingLayerSets`, which is essentially `vector<TrackingRegion, vector<SeedingLayerSet>>` i.e. for each region holds also the seeding layer pair to be used for that region.
The producer is further split in two parts
* `PixelInactiveAreaFinder`
   * Finds the inactive modules on each layer and group adjacent inactive modules together
   * Looks for layer pairs that have overlapping inactive areas and returns them and the list of layer pairs that should be used in seeding
* `AreaSeededTrackingRegionsBuilder`
   * Creates `RectangularEtaPhiTrackingRegion` for a given vector of `(phi, r, z)` boxes such that the eta-phi window covers all the boxes

As a "byproduct"
* created `AreaSeededTrackingRegionsProducer`
  * uses also `AreaSeededTrackingRegionsBuilder` to create `RectangularEtaPhiTrackingRegion`s, but the input `(phi, r, z)` boxes come from configuration
  * (actually I created this producer as one of the first steps even if it is not strictly needed for this PR, but it could be useful in the future)
* abstracted the vertex/beamspot origins to their own class
   * common part with `CandidatePointSeededTrackingRegionsProducer`
* in tracking DQM
   * monitor eta-phi areas covered by `RectangularEtaPhiTrackingRegion`s
     * currently we monitor only the direction
     * with variable-size regions monitoring also the covered area becomes interesting, and also allows to correlate regions wrt. pixel maps
   * in the configuration of "track building and seeding monitoring" replace the dictionaries with `cms.PSet`s to be able to customize them with eras
     * there is further room to reduce copy-paste and harmonize naming of PSet members with the `TrackingMonitor` parameters

~~Enabling the code in reconstruction is left to a subsequent PR. In the mean time the following commits can be used on top of this PR to enable the mitigation.
https://github.com/makortel/cmssw/compare/areaSeededRegion_step1...areaSeededRegion_step2~~

Tested in CMSSW_10_0_X_2017-11-27-2300 (rebased to CMSSW_10_0_0_pre2), ~~no changes expected~~ expecting changes in phase1 workflows as shown in the MTV plots above.

@VinInn @nikoleskinen @JanFSchulte @mtosi

makortel committed:
First version of AreaSeededTrackingRegionsProducer

makortel committed:
Enable automatic inactive regions in PixelPairStep configuration

makortel committed:
Add extraPhi and extraEta parameters

makortel committed:
Abstract producing regions from the definition of areas

makortel committed:
Add code to detect regions in pixel that are covered only by two layers

All of the useful code in PixelInactiveAreaFinder is from Niko
Leskinen (HIP/TUT, Finland). I added only some skeleton code to
finally glue Niko's code with the rest.

makortel committed:
Add skeleton of PixelInactiveAreaTrackingRegionsSeedingLayersProducer

Not all parts are ready yet, but I want to track the evolution with git.

makortel committed:
Add layer/disk to the printout

makortel committed:
Deliver layer information to PixelInactiveAreaFinder, construct mapping from active->inactive layer pairs

makortel committed:
Translate incoming active layer sets to inactive layer sets

makortel committed:
Simplify

makortel committed:
Calculate overlapping inactive areas by pairs of layers

makortel committed:
Give vertex positions to PixelInactiveAreaFinder

makortel committed:
Abstract beamspot/vertex origin to its own class

makortel committed:
Plug TrackingRegionsSeedingLayerSets into HitPairEDProducer

makortel committed:
Cleanup

- Rename member variables for clarity
- Follow member naming convention
- More functions in to anonymous namespace
- Remove unused function

makortel committed:
Extend Area to include min+max radius in order to cover FPix areas as well

makortel committed:
Deduce number of BPix ladders and number of modules per ladder from tracker geometry

makortel committed:
Read remaining sources of inactive modules

makortel committed:
Avoid unnecessary trigonometry

makortel committed:
Monitor areas covered by eta-phi tracking regions

makortel committed:
Replace bunch of dicts with dict of PSet

Psets allow era-bsed customizations. Also there is less copy-paste of
the iteration names.

makortel committed:
Monitor regions from pixelPairStep B seeding

makortel committed:
Allow ignoring inactive single-panel areas from FPix

makortel committed:
Ignore single inactive panels

makortel committed:
Add pixelPairStepTrackingRegionsSeedLayersB to timing

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21630/2395

Code check has found code style and quality issues which could be resolved by applying a patch in https://cmssdt.cern.ch/SDT/code-checks/PR-21630/2395/git-diff.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/PR-21630/2395/git-diff.patch | patch -p1`

You can run `scram build code-checks` to apply code checks directly

makortel committed:
Code checks

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21630/2401

cmsbuild commented:
A new Pull Request was created by  @makortel (Matti Kortelainen) for master.

It involves the following packages:

DQM/TrackingMonitor
DQM/TrackingMonitorSource
RecoTracker/TkHitPairs
RecoTracker/TkSeedingLayers
RecoTracker/TkTrackingRegions
TrackingTools/TransientTrackingRecHit
Validation/RecoTrack


@perrotta, @vazzolini, @kmaeshima, @dmitrijus, @cmsbuild, @jfernan2, @slava77, @vanbesien can you please review it and eventually sign? Thanks.
@ghellwig, @felicepantaleo, @abbiendi, @fioriNTU, @threus, @hdelanno, @battibass, @jhgoh, @HuguesBrun, @trocino, @GiacomoSguazzoni, @rovere, @VinInn, @bellan, @wmtford, @mschrode, @idebruyn, @ebrondol, @mtosi, @dgulhan, @calderona, @gpetruc this is something you requested to watch as well.
@davidlange6, @slava77 you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

makortel commented:
@cmsbuild, please test

I also added link to MTV plots showing the effect of this feature (when enabled) to the PR description.

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 905109682e5cf75d414b5505dd82fdd0ed9dc6dd
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21630/24914/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/bb52437ebfcd0537a73be3feed966b561f5a6709
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21630/24914/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21630/24914/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21630/24914/summary.html

There are some workflows for which there are errors in the baseline:
 10824.0 step 5
 The results for the comparisons for these workflows could be incomplete 
This means most likely that the IB is having errors in the relvals.The error does NOT come from this pull request

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 26
 - DQMHistoTests: Total histograms compared: 2606833
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2606661
 - DQMHistoTests: Total skipped: 171
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 68.8599999999 KiB( 22 files compared)
 - Checked 106 log files, 8 edm output root files, 26 DQM output files

slava77 commented:
@makortel 

are there significant doubts about enabling this code?
If the goal is to have it on by default, it may be easier to enable it from the start.

makortel commented:
@slava77 

> are there significant doubts about enabling this code?
> If the goal is to have it on by default, it may be easier to enable it from the start.

I don't think there are from TRK side. I followed @VinInn's suggestion to separate the code and enabling the feature to separate PRs.

slava77 commented:
On 12/3/17 12:05 PM, Matti Kortelainen wrote:
> @slava77 <https://github.com/slava77>
> 
>     are there significant doubts about enabling this code?
>     If the goal is to have it on by default, it may be easier to enable
>     it from the start.
> 
> I don't think there are from TRK side. I followed @VinInn
> <https://github.com/vininn>'s suggestion to separate the code and
> enabling the feature to separate PRs.

I prefer to review running code.

> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/21630#issuecomment-348810045>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEdcbovwQYvqtTFjKu-4Pmrnu85O8AOjks5s8v8AgaJpZM4Qyi60>.
>

VinInn commented:
let's enable...

cmsbuild commented:
The code-checks are being triggered in jenkins.

makortel commented:
Enabled.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21630/2410

makortel commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #21630 was updated. @perrotta, @vazzolini, @kmaeshima, @dmitrijus, @cmsbuild, @jfernan2, @slava77, @vanbesien can you please check and sign again.

cmsbuild commented:
+1
Tested at: b1c5ad2081229251dbbf04f43b555305ead7e539
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21630/24924/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/4a53445468c11d1a5fd94f68b94b9ea8be0228e9
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21630/24924/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21630/24924/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21630/24924/summary.html

There are some workflows for which there are errors in the baseline:
 10824.0 step 5
 The results for the comparisons for these workflows could be incomplete 
This means most likely that the IB is having errors in the relvals.The error does NOT come from this pull request

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /build/cmsbld/jenkins-workarea/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-21630/1325.7_TTbar_13_94XNanoAODINPUT+TTbar_13_94XNanoAODINPUT+NANOEDMMC2017+NANOMERGE

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 1524 differences found in the comparisons
 - DQMHistoTests: Total files compared: 27
 - DQMHistoTests: Total histograms compared: 2796760
 - DQMHistoTests: Total failures: 665
 - DQMHistoTests: Total nulls: 1100
 - DQMHistoTests: Total successes: 2794817
 - DQMHistoTests: Total skipped: 178
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 159.58 KiB( 23 files compared)
 - Checked 114 log files, 10 edm output root files, 27 DQM output files

slava77 commented:
On 12/4/17 12:56 AM, Matti Kortelainen wrote:
> Enabled.

In the case it needs to be disabled,
does it require applying a patch in the code with recompilation
or can we make it configurable?

> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/21630#issuecomment-348898520>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEdcbgqEkBEZ9vWu2xN1G5rh2pdQDWeRks5s87OpgaJpZM4Qyi60>.
>

makortel commented:
@slava77 

> In the case it needs to be disabled,
> does it require applying a patch in the code with recompilation
> or can we make it configurable?

In short, can be done configurable.

But what should be the expected effect? No mitigation at all, or the previous "static" mitigation? (for "no mitigation" I'd add a new boolean flag to produce empty output collections, the latter would require would require customizing the ˋPixelPairStepTaskˋ)

slava77 commented:
On 12/4/17 6:23 AM, Matti Kortelainen wrote:
> @slava77 <https://github.com/slava77>
> 
>     In the case it needs to be disabled,
>     does it require applying a patch in the code with recompilation
>     or can we make it configurable?
> 
> In short, can be done configurable.
> 
> But what should be the expected effect? No mitigation at all, or the
> previous "static" mitigation? (for "no mitigation" I'd add a new boolean
> flag to produce empty output collections, the latter would require would
> require customizing the ˋPixelPairStepTaskˋ)

I had in mind going back to the "static" mitigation
 if there is need to run alternative relvals to validate this feature alone.
That's if it's still practical to do it [I haven't looked at the
implementation]

> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/21630#issuecomment-348975892>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEdcbudD-mwgYRp9BNCL7BApU2KSQ-zZks5s9ABHgaJpZM4Qyi60>.
>

makortel commented:
> I had in mind going back to the "static" mitigation
>  if there is need to run alternative relvals to validate this feature alone.
> That's if it's still practical to do it [I haven't looked at the
> implementation]

Ok. That requires either an era switch or a customize function (but I will assume such a switch is not needed unless explicitly asked).

slava77 commented:
On 12/4/17 6:47 AM, Matti Kortelainen wrote:
>     I had in mind going back to the "static" mitigation
>     if there is need to run alternative relvals to validate this feature
>     alone.
>     That's if it's still practical to do it [I haven't looked at the
>     implementation]
> 
> Ok. That requires either an era switch or a customize function (but I
> will assume such a switch is not needed unless explicitly asked).

Please post the content of the customize function to be able to check
that this actually works

> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/21630#issuecomment-348982991>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEdcbp1THxxqnOYBKNnPuMX69MOUWDDtks5s9AXogaJpZM4Qyi60>.
>

makortel commented:
> Please post the content of the customize function to be able to check
> that this actually works

Would it be ok to push the function to git (under ˋRecoTracker/Configurationˋ)? If I craft such a function I'd prefer to store it more permanently than in a github comment.

slava77 commented:
On 12/4/17 6:58 AM, Matti Kortelainen wrote:
>     Please post the content of the customize function to be able to check
>     that this actually works
> 
> Would it be ok to push the function to git (under
> ˋRecoTracker/Configurationˋ)? If I craft such a function I'd prefer to
> store it more permanently than in a github comment.

sure, yes.

> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/21630#issuecomment-348986590>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEdcbgbQVQeMu_m4HVHX9C-Vvhekpen2ks5s9AifgaJpZM4Qyi60>.
>

makortel committed:
Add customize function to switch back to 'static' mitigation

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21630/2425

cmsbuild commented:
Pull request #21630 was updated. @perrotta, @vazzolini, @kmaeshima, @dmitrijus, @cmsbuild, @jfernan2, @slava77, @vanbesien can you please check and sign again.

makortel commented:
Added customize function in `RecoTracker/Configuration/python/customisePixelPairStaticMitigation.py`. I tested that RECO step runs with it and that the reco configuration (via `edmConfigDump`) is the same as with 905109682e5cf75d414b5505dd82fdd0ed9dc6dd (DQM+validation won't be exactly the same because that would be much easier done with era so for the possible foreseen use case I didn't bother).

makortel commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 99315fdaffa5194ceee0139e67b680ab6a0e7a09
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21630/24939/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/632b6c042baf87d350cb619157835bedf44e4be0
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21630/24939/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21630/24939/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21630/24939/summary.html

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /build/cmsbld/jenkins-workarea/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-21630/1325.7_TTbar_13_94XNanoAODINPUT+TTbar_13_94XNanoAODINPUT+NANOEDMMC2017

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 1524 differences found in the comparisons
 - DQMHistoTests: Total files compared: 27
 - DQMHistoTests: Total histograms compared: 2835085
 - DQMHistoTests: Total failures: 881
 - DQMHistoTests: Total nulls: 1359
 - DQMHistoTests: Total successes: 2832667
 - DQMHistoTests: Total skipped: 178
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 160.04 KiB( 23 files compared)
 - Checked 113 log files, 9 edm output root files, 27 DQM output files

slava77 commented on RecoTracker/TkSeedingLayers/interface/SeedingLayerSetsBuilder.h:
```diff
@@ -32,9 +34,15 @@ class SeedingLayerSetsBuilder {
+
+  const std::vector<SeedingLayerSetsHits::LayerSetIndex>& layerSetIndices() const { return theLayerSetIndices; }
+
+  std::vector<std::vector<SeedingLayerId> > layerSets() const; // please call at most once per job per client
```
where is this implemented?

makortel replied:
Likely leftover from the prototyping phase (in the end `layers()` was more useful), will remove.

slava77 commented on RecoTracker/TkTrackingRegions/plugins/AreaSeededTrackingRegionsBuilder.h:
```diff
@@ -0,0 +1,108 @@
+    // ordered by which way one goes around the unit circle, so it may
+    // happen that actually phimax < phimin
+    Area(double rmin, double rmax, double phimin, double phimax, double zmin, double zmax):
+      m_zmin(zmin), m_zmax(zmax)
```
is double precision required here?

makortel replied:
Very likely not, especially given that all the subsequent calculations in `AreaSeededTrackingRegionsBuilder` are done in float. I'll change these to `float`.

slava77 commented on RecoTracker/TkTrackingRegions/plugins/AreaSeededTrackingRegionsBuilder.cc:
```diff
@@ -0,0 +1,245 @@
+      // decide which phi is the largest/smallest without having too
+      // much of headache with the wrapping
+      const auto phi_ref = std::atan2(0.5f*(y_rmin_phimin + y_rmin_phimax),
+                                      0.5f*(x_rmin_phimin + y_rmin_phimax));
```
should the second argument be with *x*? ``` 0.5f*(x_rmin_phimin + x_rmin_phimax) ```.
Also, 0.5f appears to be redundant here.

slava77 replied:
Before I spend more time looking at the comparisons, please clarify if this is a bug that would affect performance.

makortel replied:
I agree, thanks for catching (I'm running MTV to test the effect).

makortel replied:
It can affect, so better to wait for an update (hopefully later today).

slava77 commented on RecoTracker/TkTrackingRegions/plugins/AreaSeededTrackingRegionsProducer.h:
```diff
@@ -0,0 +1,118 @@
+#ifndef RecoTracker_TkTrackingRegions_AreaSeededTrackingRegionsProducer_h
```
[minor/aside remark] "Generator" or "Maker" in the name may be a good disambiguation from mistaking this class name for an EDProducer.

makortel replied:
In general I agree, but a change in the naming convention (originating from 2007...) should be considered separately from this PR. On the other hand the relatively recent change of producing `TrackingRegions` to `Event` increased the possibilities for confusion.

slava77 commented on RecoTracker/TkTrackingRegions/plugins/AreaSeededTrackingRegionsProducer.h:
```diff
@@ -0,0 +1,118 @@
+    std::vector<edm::ParameterSet> vDefaults;
+    /*
+    edm::ParameterSet vDefaults1;
+    vDefaults1.addParameter<double>("rmin", 0.0);
```
is this commented out code needed?
Please remove or add comments inline in the code why the commented out block is relevant

makortel replied:
No, will remove.

slava77 commented on RecoTracker/TkTrackingRegions/plugins/PixelInactiveAreaFinder.cc:
```diff
@@ -0,0 +1,1104 @@
+        //Stream ss;
+        //detGroupSpanInfo(cspanA,ss);ss<<std::endl;detGroupSpanInfo(cspanB,ss);ss<<std::endl;
+        //edm::LogPrint("") << ss.str();ss.str(std::string());
+        //edm::LogPrint("") << "**";
```
is this commented out code needed?
Please remove or add comments inline in the code why the commented out block is relevant

slava77 replied:
If this and the other printouts are actually needed, better wrap them in EDM_ML_DEBUG and uncomment.

makortel replied:
The printouts look like they could be useful also later, so I'll wrap them to EDM_ML_DEBUG.

slava77 commented on RecoTracker/TkTrackingRegions/plugins/PixelInactiveAreaFinder.cc:
```diff
@@ -0,0 +1,1104 @@
+            detGroupSpanInfo(outerSpan, ss);
+            LogTrace("PixelInactiveAreaFinder") << " adding areas for active layer sets " << ss.str();
+#endif
+            ret.emplace_back(std::vector<DetGroupSpan>{{innerSpan, outerSpan}}, (*layerSetIndexInactiveToActive_)[i]);
```
here and for areasAndLayerSets the first argument is size 2. Can it be made to an array [or another fixed size selection] instead of a vector?
... ignore this if there is another modification that can extend this past 2

makortel replied:
I had this in mind as well but apparently forgot. I don't think there is any apparent need for anything beyond 2 (as that would mean we would try to mitigate cases with >= 3 inactive layers; ok possible for phase2 FPix but I would ignore that for now). Size 1 could be useful for "triplet mitigation for HLT" planned by @JanFSchulte, so I'll make it `edm::VecArray` to have consistent interface.

makortel committed:
Remove unneeded function declaration

makortel committed:
Remove unnecessary commented code

makortel committed:
Wrap commented printouts in EDM_ML_DEBUG, set message category everywhere

makortel committed:
Change arguments to float

makortel committed:
Use VecArray for the inactive area coordinates

There are currently always two of them, later it may "extend" to 1
(for "triplet mitigation" for HLT), which is why I used VecArray
instead of std::array.

makortel committed:
Fix a bug

(x_rmin_phimin+y_rmin_phimax) -> (x_rmin_phimin+y_rmin_phimax)

cmsbuild commented:
The code-checks are being triggered in jenkins.

makortel committed:
Code checks

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
The tests are being triggered in jenkins.
https://cmssdt.cern.ch/jenkins/job/ib-any-integration/25062/console Started: 2017/12/12 20:07

cmsbuild commented:
+1
Tested at: 39f3ff2c6d1693e249ff4c70ca1198aec6ed836a
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21630/25062/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/e2188ba545ca2f19834bf458d40003ea8d11f10c
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21630/25062/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21630/25062/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21630/25062/summary.html

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /build/cmsbld/jenkins-workarea/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-21630/1325.7_TTbar_13_94XNanoAODINPUT+TTbar_13_94XNanoAODINPUT+NANOEDMMC2017

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 1529 differences found in the comparisons
 - DQMHistoTests: Total files compared: 27
 - DQMHistoTests: Total histograms compared: 2835241
 - DQMHistoTests: Total failures: 890
 - DQMHistoTests: Total nulls: 1356
 - DQMHistoTests: Total successes: 2832817
 - DQMHistoTests: Total skipped: 178
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 159.76 KiB( 22 files compared)
 - Checked 113 log files, 9 edm output root files, 27 DQM output files

dmitrijus commented:
+1

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/21630	39f3ff2
- changes are in line with the PR description: aiming to improve pixel pair recovery by making regions for pixelPairStep based on the information known from the conditions or bad ROC data event by event. This affects only phase 1.
- jenkins tests pass and comparisons with baseline show somewhat small changes in phase-1 workflows starting from tracking variables and propagating downstream
- local tests confirm expected behavior

at the track level there is some increase in efficiency in bad regions of eta/phi
<img width="797" alt="wfmatti_ttbarpu50_efffake_eta" src="https://user-images.githubusercontent.com/4676718/33967026-9e8c6dc8-e016-11e7-8558-2f55a1e9fbde.png">

the effect is visible more clearly in a run with a larger fraction of bad modules, e.g. 305064 (wf 136.829
In this run the bad regions can be seen in the z-phi or xy plots (picked from offline DQM)
<img width="797" alt="run305064_dqm_clupos_bpix_zphi" src="https://user-images.githubusercontent.com/4676718/33967103-ecfc7ff2-e016-11e7-9215-d65512ea0c3e.png">

<img width="797" alt="run305064_dqm_clupos_fpix_zphi" src="https://user-images.githubusercontent.com/4676718/33967120-f8e2f6ca-e016-11e7-9362-9d4f39566149.png">

Here the pixelPair seeding with this PR is more "selective" around the partially dead regions
(left is baseline, right is this PR)
<img width="875" alt="wf136 829_pixpair_seeds_etaphi" src="https://user-images.githubusercontent.com/4676718/33967169-298e80b4-e017-11e7-930f-13463ded36f9.png">
this eventually propagates to the generalTracks level with about 10% increase in the pixelPair algo tracks and even some decrease in pixelLess
![all_sign984vsorig_rundoubleeg2017fwf136p829c_recotracks_generaltracks__rereco_obj_algo](https://user-images.githubusercontent.com/4676718/33967339-e1be26da-e017-11e7-9607-31d6a8dc44d5.png)

High level quantities are not affected significantly, mostly because the effect is rather small.

CPU per event in reco has decreased by about 1-2% in 10224 (somewhat significant above general measurement uncertainty) and no clear change in 2017F data workflows in (up by about 1% in 136.829/DoubleEG and down by about 1% in 136.83/DoubleMuon).

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

JanFSchulte commented:
For completeness sake: This new recovery procedure has also been tested, see the report at the TSG workshop yesterday: https://indico.cern.ch/event/674023/contributions/2805626/attachments/1574212/2485184/20171212_TSGWorkshop_Tracking.pdf

It certainly has the potential to recover a lot of lost efficiency there, but the procedure needs a bit of refining going forward to keep the timing under control.

davidlange6 commented:
+1
